pkg/stub/stub.go

```diff
@@ -441,22 +452,33 @@ func (stub *stub) Run(ctx context.Context) error {
 		return err
 	}
 
-	err = <-stub.srvErrC
-	if err == ttrpc.ErrServerClosed {
-		return nil
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err = <-stub.srvErrC:
+			if isRecoverableErr(err) {
+				log.Warnf(ctx, "Plugin service stopped", "error", err)
+				continue
+			}
+			return err
+		}
 	}
+}
```

Comment:
I don't think this is a good idea. The plugin should be considered running while it is serving requests. Once it stops serving requests it should not be considered running, so `Run()` should return.

This looks like an attempt at a variation of your original scheme where you tried to make the stub internally restart itself. Now the stub tries to hide the fact that it is not running any more.

Instead what we should do IMO, and I think this is also what we discussed earlier, is to make the stub re-`Start()`-able. IOW, change it so, that calling `stub.Stop()` cleans up/resets the internal state of the plugin so, that a `stub.Start()` succeeds again (for externally launched plugins, for others this cannot be done).